HTS Design Doc (#1061)
hashgraph/hedera-mirror-node: docs/design/hts.md

# Hedera Token Service Design

## Purpose

The Hedera Token Service (HTS) builds upon the Cryptocurrency Service to provide decentralized issuance of custom tokens on the Hedera Network.
The behavior will be similar to that of the native HBAR token and as such the Mirror Node will persist token balances and transfer lists and support the retrieval of information through its APIs.

This document highlights the architecture and design changes to be made to support HTS.

> Changes should be applied in order of the following order to support easy third party integration.

Author: Fixed

## Architecture

![Architecture](images/hts-architecture.png)

1. Downloader retrieves transactions and balances (HBAR & Tokens) from stream files in cloud bucket and validates them
2. Parser persists to database
3. Client queries balance or token APIs for details
4. REST API retrieves token details and balances from database
5. Client subscribes to GRPC API for token transfers
6. GRPC API retrieves token transfers from database and returns to client
7. GRPC API is notified of new token transfers via database streaming logic

## Alternatives

## Database
To support the goals the following database schema changes should be made
1. `t_entity_types` updated to represent new `TOKEN` entity
2. `t_transaction_types` updated to handle token transactions from network.
3. New `token_balance` table added to persist token only balances
4. New `token` table added to capture token specific entity properties
5. New `token_transfer` table to persist token transfers
6. `t_transaction_results` updated to handle token transactions results from network.
7.  New `token_account` table added to persist token specific account state

### Token Transfer

> -   Add `token_transfer` table table. with the following class members

Author: Fixed

2. `t_transaction_types` updated to handle token transactions from network.
3. New `token_balance` table added to persist token only balances
4. New `token` table added to capture token specific entity properties
5. New `token_transfer` table to persist token transfers
6. `t_transaction_results` updated to handle token transactions results from network.
7.  New `token_account` table added to persist token specific account state

### Token Transfer
-   Add `token_transfer` table table. with the following class members
    -   `account_id`
    -   `amount`
    -   `consensus_timestamp`
    -   `token_id`
-   Add a `token_transfer__token_account_timestamp` index with `desc` order on columns

### Token Balance
-   Create `token_balance` table to distinctly capture token balances with the following columns
    -   `consensus_timestamp`
    -   `account_id`
    -   `balance`
    -   `token_id`
-   Create primary key as combination of `consensus_timestamp`, `token_id` and `account_id`

### Token Account Info
-   Create `token_account` table to distinctly capture token account metadata changes with the following columns
    -   `id` (primary key)
    -   `account_id`

>     -   `create_timestamp`

Author: Was changed to 
```suggestion
    -   `created_timestamp`
```

    -   `frozen`
    -   `kyc`

>     -   `modify_timestamp`

Author: Was changed to
```suggestion
    -   `modified_timestamp`
```

        (182, 'TOKEN_HAS_NO_SUPPLY_KEY'),
        (183, 'TOKEN_HAS_NO_WIPE_KEY'),
        (184, 'INVALID_TOKEN_MINT_AMOUNT'),
        (185, 'INVALID_TOKEN_BURN_AMOUNT'),
        (186, 'ACCOUNT_HAS_NO_TOKEN_RELATIONSHIP'),
        (187, 'CANNOT_WIPE_TOKEN_TREASURY_ACCOUNT'),
        (188, 'INVALID_KYC_KEY'),
        (189, 'INVALID_WIPE_KEY'),
        (190, 'INVALID_FREEZE_KEY'),
        (191, 'INVALID_SUPPLY_KEY'),
        (192, 'TOKEN_NAME_ALREADY_IN_USE'),
        (193, 'MISSING_TOKEN_NAME'),
        (194, 'TOKEN_NAME_TOO_LONG'),
        (195, 'INVALID_WIPING_AMOUNT'),
        (196, 'TOKEN_IS_IMMUTABlE'),
        (197, 'TOKEN_ALREADY_ASSOCIATED_TO_ACCOUNT');
```

### Entities
-   Insert token entity info into `t_entities` table from `TokenCreation.proto` transaction object. Insert `adminKey` as `key`, `expiry` as `exp_time_ns`, `autoRenewAccount` as `auto_renew_account_id` and `autoRenewPeriod` as `auto_renew_period`.

### Token
-   Create `token` table with the following columns. Table will capture non shared `entity` items, most API calls may not require this information and therefore additional sql joins may be avoided.
    -   `token_id` (primary key)

>     -   `create_timestamp`

Author: Was changed to
```suggestion
    -   `created_timestamp`
```

    -   `divisibility`
    -   `freeze_default`
    -   `freeze_key`
    -   `initial_supply`
    -   `kyc_default`
    -   `kyc_key`

>     -   `modify_timestamp`

Author: Was changed to
```suggestion
    -   `modified_timestamp`
```

    -   `supply_key`
    -   `symbol`
    -   `treasury_account_id`
    -   `wipe_key`

## Importer

### Converter
-   Add a `TokenIdConverter`
```java
    package com.hedera.mirror.importer.converter;
    ...
    public class TokenIdConverter extends AbstractEntityIdConverter {

        public TokenIdConverter() {
            super(EntityTypeEnum.TOKEN);
        }
    }
```

### Domain

#### AccountBalance
-   Update `AccountBalance` to have a `tokenBalances` class member, to represent all the token balances of the account
```java
public class AccountBalance implements Persistable<AccountBalance.Id> {
    ...
    private List<TokenBalance> tokenBalances;
    ...
}
```

#### TokenTransfer
-   Create new `TokenTransfer` with the following class members
    -   `accountId`
    -   `amount`
    -   `consensusTimestamp`
    -   `tokenId`

#### EntityTypeEnum
-   Update `EntityTypeEnum` with `Token` type
```java
    public enum EntityTypeEnum {
        ...
        TOKEN(5);
        ...
    }
```

#### Token
-   Add `Token` class to hold Token specific metadata outside of the base entity class with the following class members

>     -   `createTimestamp`

Author: ```suggestion
    -   `createdTimestamp`
```

    -   `divisibility`
    -   `freezeDefault`
    -   `freezeKey`
    -   `initialSupply`
    -   `kycDefault`
    -   `kycKey`

>     -   `modifyTimestamp`

Author: ```suggestion
    -   `modifiedTimestamp`
```

    -   `supplyKey`
    -   `symbol`
    -   `tokenId`
    -   `treasuryAccountId`
    -   `wipeKey`

#### TokenAccount
-   Add `TokenAccount` class with the following class members
    -   `id`
    -   `accountId`

>     -   `createTimestamp`

Author: ```suggestion
    -   `createdTimestamp`
```

    -   `frozen`
    -   `kyc`

>     -   `modifyTimestamp`

Author: ```suggestion
    -   `modifiedTimestamp`
```

    -   `tokenId`
    -   `wiped`

#### TokenBalance
-   Add `TokenBalance` class with the following class members
    -   `accountId`
    -   `balance`
    -   `consensusTimestamp`
    -   `tokenId`

### Balance Parsing
To support HTS the balance CSV
1. Adds a version comment as the first line of the file i.e. `# version:2`

> 2. Comments out the Timestamp e.g. `TimeStamp:2020-09-22T04:25:00.083212003Z`

Author: ```suggestion
2. Comments out the Timestamp e.g. `# TimeStamp:2020-09-22T04:25:00.083212003Z`
```

3. Adds a `tokenBalances` column which is the Base64 encoding of the serialized bytes of `TokenBalances` proto.

To allow the mirror node to support both V1 and V2 balance files
-   Rename `BalanceFileReaderImpl` to `BalanceFileReaderImplV1`
-   Add a `BalanceFileReaderImplV2` class to parse `shardNum,realmNum,accountNum,balance,tokenBalances`.  Reader should set `tokenBalances` in `AccountBalance`.

> -   Add a `CompositeBalanceFileReader` that implements `BalanceFileReader` and reads the first line of the CSV

Reviewer: Do we still want the parser to be lenient on the number of lines before the timestamp/version number?  Looks like right now allow the timestamp to be up to 10 lines deep, so long as the column headers don't come up first.

Author: Yeah I'd see leave that lenience for V1. I'm presuming it's there for padded cases? 
We should verify why.
Given the file is now versioned it would be the best time to dump that logic for V2 if it's no longer needed.

-   `TransactionBody.hasTokenUpdate()` and parse `TokenUpdateTransactionBody` out from the record. Retrieve an existing `Token` db entry, update the appropriate columns and pass it to `entityListener.onToken()`.

## REST API
To achieve the goals and for easy integration with existing users the REST API should be updated in the following order
1.  The `transactions` REST API must be updated to support `tokenTransfers`
2.  The `balances` REST API must be updated to support `tokenBalances`
3.  The `accounts` REST API must be updated to support `tokenBalances`
4.  Add a Token Supply distribution REST API to show token distribution across accounts - `/api/v1/tokens/<symbol>/balances`
5.  Add a Token Discovery REST API to show available tokens on the network - `/api/v1/tokens`
6.  Add a Token Info REST API to show details for a token on the network - `/api/v1/tokens/<symbol>`

### Accounts Endpoint
-   Update `/api/v1/accounts` response to add token balances
```json
{
    "accounts": [
      {
        "balance": {
          "timestamp": "0.000002345",
          "balance": 80,
          "token_balances": [
            {

>               "symbol": "FOOBAR",

Author: Will be changed to tokenId
```suggestion
              "token_id": "0.1.2",
```

              "balance": 80
            },
            {

>               "symbol": "FOOCOIN",

Author: ```suggestion
              "token_id": "0.3.4",
```

                "balance": 80
              },
              {
                "symbol": "FOOCOIN",
                "balance": 50
              }
            ]
          }
        ],
        "links": {
          "next": null
        }
    }
```

### Transactions Endpoint
-   Update `/api/v1/transactions` response to add token transfers
```json
    {
      "transactions": [
        {
          "consensus_timestamp": "1234567890.000000001",
          "valid_start_timestamp": "1234567890.000000000",
          "charged_tx_fee": 7,
          "memo_base64": null,
          "result": "SUCCESS",
          "transaction_hash": "aGFzaA==",
          "name": "CRYPTOTRANSFER",
          "node": "0.0.3",
          "transaction_id": "0.0.10-1234567890-000000000",
          "valid_duration_seconds": "11",
          "max_fee": "33",
          "transfers": [
            {
              "account": "0.0.9",
              "amount": 10
            },
            {
              "account": "0.0.10",
              "amount": -11
            },
            {
              "account": "0.0.98",
              "amount": 1
            }
          ],
          "token_transfers": [
            {

>               "transfers": [

Reviewer 2: Spurious transfers key and extra array

Author: The idea was to have it similar to what was noted in the design doc copied below
```json
{
    transfers: [ 
        {account: 0.0.1234, amount: -1000},  
        {account: 0.0.98, amount: 1000} 
    ], 
    tokenTransfers: [ 
        {
            token: 0.0.1234, 
            transfers: [ 
                {account: 0.0.1234, amount: -1000},  
                {account: 0.0.98, amount: 1000} 
            ]
        
        }
    } 
}
```

However I left out the aggregation of the token. Corrected the `token_transfers` object should look like this
```suggestion
              {
    "token_id": "0.15.3",
    "transfers": [
        {
            "account": "0.0.1111",
            "amount": -10,        
        },
        {
            "account": "0.0.2222",
            "amount": 10
        }
    ]
},
{
    "token_id":"0.2.4"
    "transfers": [
        {
            "account": "0.0.3333",
            "amount": -10
        },
        {
            "account": "0.0.4444",
            "amount": 10
        }
    ]
}
```

Author: Looks like it is possible to transfers span multiple tokens in a transaction, so grouping on a token would be useful

Reviewer 2: We had said before in a previous [comment](https://github.com/hashgraph/hedera-mirror-node/pull/1061#discussion_r493192465) that we were okay with the duplication since it matches the database and proto structure more.

Author: Yep, thanks for circling back. Will change to 
```suggestion
              "token_transfers": [
    {
        "token_id": "0.15.3",
        "account": "0.0.1111",
        "amount": -10
    },
    {
        "token_id": "0.15.3",
        "account": "0.0.2222",
        "amount": -10
    },
    {
        "token_id": "0.2.4",
        "account": "0.0.3333",
        "amount": -10
    },
    {
        "token_id": "0.2.4",
        "account": "0.0.4444",
        "amount": -10
    }
]
```

                {
                  "account": "0.0.1111",
                  "amount": -10,
                  "symbol": "FOOBAR"
                },
                {
                  "account": "0.0.2222",
                  "amount": 10,
                  "symbol": "FOOBAR"
                }
              ]
            },
            {
              "transfers": [
                {
                  "account": "0.0.3333",
                  "amount": -10,
                  "symbol":"FOOCOIN"
                },
                {
                  "account": "0.0.4444",
                  "amount": 10,
                  "symbol":"FOOCOIN"
                }
               ]
            }
          ]
        }
      ]
    }
```

### Token Supply distribution

> `/api/v1/tokens/<symbol>/balances` this could be the equivalent of `/api/v1/balances?symbol=<symbol>` currently and would return a list of account and the token balances

Author: ```suggestion
`/api/v1/tokens/<tokenId>/balances` this could be the equivalent of `/api/v1/balances?account.id=<account.id>` currently and would return a list of account and the token balances
```

```json
    {
        "timestamp": "0.000002345",
        "balances": [
          {
            "account": "0.15.10",
            "balance": 100
          },
          {
            "account": "0.15.9",
            "balance": 90
          },
          {
            "account": "0.15.8",
            "balance": 80
          }
        ],
        "links": {
          "next": null
        }
    }
```

Optional Filters
-   `/api/v1/tokens/<symbol>/balances?id=0.0.1000`
-   `/api/v1/tokens/<symbol>/balances?balance=gt:1000`

> -   `/api/v1/tokens/<symbol>/balances?timestamp=1566562500.040961001`

Author: Additionally add publickey to match token adminKey
```suggestion
-   `/api/v1/tokens/<tokenId>/balances?timestamp=1566562500.040961001`
-   `/api/v1/tokens/< tokenId >/balances?publicKey=2b60955bcbf0cf5e9ea880b52e5b63f664b08edf6ed 15e301049517438d61864`
```
Basically we want similar parity with https://docs.hedera.com/guides/docs/mirror-node-api/cryptocurrency-api#additional-examples

### Token Discovery
`/api/v1/tokens` this would return all tokens present on the network

```json
    {
        "tokens": [
          {
            "symbol": "FOOBAR",
            "token_id": "0.15.10",
            "admin_key": {
              "_type": "ProtobufEncoded",
              "key": "7b2233222c2233222c2233227d"
            }
          },
          {
            "symbol": "FOOCOIN",
            "tokenId": "0.15.11",
            "admin_key": {
              "_type": "ProtobufEncoded",
              "key": "9c2233222c2233222c2233227d"
            }
          }
        ],
        "links": {
          "next": null
        }
    }
```

Optional Filters
-   `/api/v1/tokens?publickey=3c3d546321ff6f63d701d2ec5c277095874e19f4a235bee1e6bb19258bf362be` - All tokens with matching admin key

> -   `/api/v1/tokens?account.id=0.0.8` - All tokens for matching account

Author: Add tokenId range
```suggestion
-   `/api/v1/tokens?token.id=0.0.8` - All tokens for matching account
-   `/api/v1/tokens?token.id=gt:0.0.1001`
```

### Token Info

> -   Add a `getTokenInfo()` to `tokens.js` to retrieve token info from `token` table

Reviewer 2: Missing URL format. Can probably remove this code related line.

Author: Added and removed

The GRPC module currently supports listen and retrieval logic for incoming and historic topic messages respectively.
A lot of this logic can be shared for the equivalent `TokenTransfer` listeners and retriever

TBD

### Token Transfer Service
TBD
> _Note:_ ***Sequence diagram to be added

## Non-Functional Requirements

## Open Questions
-   [x] What's the maximum character size of the token `symbol` string
    -   A: Max length is currently 100 chars to match memo, https://github.com/hashgraph/hedera-services/blob/separate-tokenrels-fcm/hedera-node/src/main/resources/bootstrap.properties#L56
-   [x] Will a `token_id` and `token` be assigned a default value for HBARs across the network e.g. i.e. '1' and 'HBAR' respectively
    -   A: Currently no since hbar is not treated as an entity like tokens will be.
-   [x] Should token only entity items exist in their own table or be added to `t_entities`?
    -   A: Will split out into `token` table
-   [x] Should `account_balance` `accountNum` and `accountRealmNum` be migrated into `entityId` or should token_balance also use `accountNum` and `accountRealmNum` instead of `entityId`?
    -   A: Should be migrated to use `account_id` only

> -   [ ] What filter options should be provided for new Token API's

Author: Provide parity with /accounts and /balances APIs where applicable i.e. filtering on accountId, tokenId and publicKey

-   [x] How should frozen account and kyc'd account for a token be represented?
    -   A: As columns in `token` table

> -   [ ] Should balance returns for `accounts`, `balances` and `transactions` contain `token` or `tokenId` or both?

Author: Returns should be `tokenId`.
Token API would be the place where both can be returned

# Hedera Token Service Design

## Purpose

The Hedera Token Service (HTS) builds upon the Cryptocurrency Service to provide decentralized issuance of custom tokens on the Hedera Network.

> The behavior will be similar to that of the native HBAR token and as such the Mirror Node will persist token balances and transfer lists and support the retrieval of information through the API's.

Reviewer 2: nit: `through its APIs`

> This document highlights the architecture and design changes to be made on top of v0.19.0 code to support HTS.

Reviewer 2: Let's not mention specific versions as things can always change.

2. Parser persists to database
3. Client queries balance or token APIs for details
4. REST API retrieves token details and balances from database
5. Client subscribes to GRPC API for token transfers
6. GRPC API retrieves token transfers from database and returns to client
7. GRPC API is notified of new token transfers via database streaming logic

## Alternatives

## Database
To support the goals the following schema changes should be made
1. `t_entities` and ` t_entity_types` need to be update to represent new TOKEN entity
2. New `token_balance` table is added to persist token only balances
3. New `token` table is added to capture token specific entity items
4. `crypto_transfer` table should be updated to persist token transfers and also distinguish between HBAR and other token transfers
5. Update transaction types and results to handle token transactions from network.

### Crypto Transfer
-   Add columns to `crypto_transfer` table. Symbol can be pulled from a `TransactionReceipt#tokenId` of transaction
```sql
    alter table if exists crypto_transfer
        add column token_id entity_id

>         add column symbol   character varying(96) not null;

Reviewer 2: We shouldn't duplicate storage of the symbol especially in our biggest table. Also, I think token ID is immutable but token symbol is mutable. token_id should be enough.

Author: Fair point on the size of the table and mutability of the mentioned columns. 
Like above the hope was to avoid a join with another table on API queries.
Especially since eventually we'll be streaming updates from this table. Joins would be a performance hit on speed mores than the size of the table

Author: We could work out not having the symbol here.
On the `/api/v1/transactions` response the current proposed solution returns the tokenId, however there's a chance we actually want to return the symbol.
This would require a join with the token table if so.

On the gRPC API streaming side it's possible for us to make the mapping call once at the beginning for use so that's solvable.
For REST API we may have to explore caching the mapping to optimize on performance

Reviewer 2: We definitely can't have mutable data in an insert only table. I don't think the join is an issue since tokens will have thousands _maybe_ hundreds of thousands while cryptotransfer has hundreds of millions. I'm way more concerned about the mutability and storage size of a varchar field than joining with a small table. Plus we can always cache the tokenid to symbol mapping

Author: Seems like we're agreed on caching, dropping `symbol` for the mutability reasons you rightly pointed out

```
-   Add migration logic to insert default "HBAR" for `symbol` column in `crypto_transfer` for historic data

### Token Balance
-   Create `token_balance` table to distinctly capture token balances vs HBAR balances

```sql
    create table if not exists token_balance
        consensus_timestamp bigint      primary key not null,

>         symbol              character   varying(96),

Reviewer 2: Shouldn't duplicate symbol. token_id is enough.

Author: Token API's will likely query via the symbol. 
Repeating it here avoids a join with the `token` table

Author: Performance of the REST API's isn't an issue and if it becomes one we could explore caching the symbol-tokenId on the REST API side.
That's one other option if we feel strongly against the duplicate info here

Author: Dropping `symbol`

        (177, 'INVALID_TOKEN_REF'),
        (178, 'ACCOUNT_KYC_NOT_GRANTED_FOR_TOKEN'),
        (179, 'TOKEN_HAS_NO_KYC_KEY'),
        (180, 'INSUFFICIENT_TOKEN_BALANCE'),
        (181, 'TOKEN_WAS_DELETED'),
        (182, 'TOKEN_HAS_NO_SUPPLY_KEY'),
        (183, 'TOKEN_HAS_NO_WIPE_KEY'),
        (184, 'INVALID_TOKEN_MINT_AMOUNT'),
        (185, 'INVALID_TOKEN_BURN_AMOUNT'),
        (186, 'ACCOUNT_HAS_NO_TOKEN_RELATIONSHIP')
        (187, 'CANNOT_WIPE_TOKEN_TREASURY_ACCOUNT')
        (188, 'INVALID_KYC_KEY')
        (189, 'INVALID_WIPE_KEY')
        (190, 'INVALID_FREEZE_KEY')
        (191, 'INVALID_SUPPLY_KEY')
        (192, 'INVALID_TOKEN_EXPIRY')
        (193, 'TOKEN_HAS_EXPIRED')
        (194, 'TOKEN_IS_IMMUTABlE');
```

### Entities
-   Update `t_entities` table with token entity info, from the `TokenCreation.proto` object insert `adminKey` as `key`, `expiry` as `exp_time_ns`, `autoRenewAccount` as `auto_renew_account_id` and `autoRenewPeriod` as `auto_renew_period`.

### Token
-   Create `token` class to split out the non shared entity items into a new table. Most API calls may not require this information and therefore additional sql joins may be avoided.
```sql

>     create table if not exists token

Reviewer 2: Would prefer column names sorted since any other order would be arbitrary.

Author: Can do

Author: All columns after primary key are sorted a-z

        consensus_timestamp bigint  primary key not null,

>         symbol              character varying(96),

Reviewer 2: Have we confirmed with services if there's any limit imposed on the symbol size? There probably should be a limit since it's stored in state.

Author: Noted in the `Open Questions` intending to circle back with services on this and others, thanks for highlighting

Author: Current limit seems to be 32. But it's configurable so maybe we should add some padding incase of marginal increases over time

Author: Actually the [default limit](https://github.com/hashgraph/hedera-services/blob/separate-tokenrels-fcm/hedera-node/src/main/resources/bootstrap.properties#L56) for the tokens.maxTokenNameLength property is 100 UTF-8 characters. This matches the maximum length of a memo

>         token_id            entity_id,

Reviewer 2: Should token_id be the primary key? It depends on how queries will use it. Definitely token_id should have a unique index on it.

Author: Correct, good catch. token_id should be the primary key. Will change

        initial_supply      bigint,
        divisibility        bigint,

>         treasury            entity_id,

Reviewer 2: `treasury_account` or `treasury_account_id`

Author: Yes had made the change to  `treasury_account` locally, will push on next comment

Author: Actually had it as `treasury_id`

        kyc_key             bytea,
        freeze_key          bytea,
        wipe_key            bytea,

>         supply_key          bytea,

Reviewer 2: Are any of these keys equivalent to the submit_key that would make sense to reuse that field?

Author: Nope these are all different than the submit key. 
Each one allows a specific Token/Token-Account altering operation and not necessarily whether the account holder of the key can perform token transfers which I think would be the equivalent of submitKey (only those keys could submit messages to the topic entity).
They are similar in that one those keys can perform those operations

        freeze_default      boolean,
        kyc_default         boolean,

>         freeze_default      boolean;

Author: remove repeated
```suggestion
;
```

-   Add `TokenBalanceSerializer` converter to handle object to JSON string serialization
```java
   @Named
   public class TokenBalanceSerializer extends JsonSerializer<List<TokenBalance>> {
       @Override
       public void serialize(List<TokenBalance> tokens, JsonGenerator gen, SerializerProvider serializers) throws IOException {
           if (value != null) {
               gen.writeStartArray();
               for (TokenBalance token: tokens) {
                   gen.writeStartObject();
                   gen.writeObjectField("token", token);
                   gen.writeEndObject();
               }
               gen.writeEndArray();
           }
       }
   }
   ```

or

```java
    @Named
    public class TokenBalanceSerializer extends JsonSerializer<List<TokenBalance>> {
        @Override
        public void serialize(List<TokenBalance> tokens, JsonGenerator gen, SerializerProvider serializers) throws IOException {
            if (value != null) {
                gen.writeStartArray();
                for (TokenBalance token: tokens) {
                    gen.writeObject(token);
                }
                gen.writeEndArray();
            }
        }
    }

> ```

Author: Ignore TokenBalanceSerializer, left over from a previous thought, will remove

### Converter
-   Add a `TokenIdConverter`
```java
    package com.hedera.mirror.importer.converter;
    ...
    public class TokenIdConverter extends AbstractEntityIdConverter {

        public FileIdConverter() {
            super(EntityTypeEnum.TOKEN);
        }
    }
```

### Domain

> -   Add `TokenBalance` with `tokens` private class member

Author: ```suggestion
-   Add `TokenBalance`
```

```java
    public class TokenBalance implements Persistable<TokenBalance.Id> {

        private long balance;

        private String symbol;

        @EmbeddedId
        @JsonUnwrapped
        private Id id;

        @Override
        public boolean isNew() {
            return true; // Since we never update balances and use a natural ID, avoid Hibernate querying before insert
        }

        @Data
        @AllArgsConstructor
        @NoArgsConstructor
        @Embeddable
        public static class Id implements Serializable {

>             private static final long serialVersionUID = -2399552489266593375L;

Author: replace serialVersionUID w unique one as this was copied

            private EntityId tokenId;

            @Convert(converter = AccountIdConverter.class)
            @JsonSerialize(using = EntityIdSerializer.class)
            private EntityId accountId;
        }
    }
```

-   Update `CryptoTransfer` to have a `token_id` and `symbol` class members, to allow it to represent both HBARs and Tokens

-   Add `Token` class to hold Token specific metadata outside of the base entity
```java
    public class Token {
        private EntityId tokenId;
        private String symbol;
        private Long initialSupply;
        private Long  divisibility;
        @Convert(converter = AccountIdConverter.class)
        private EntityId treasury;
        private byte[] kycKey;
        private byte[] freezeKey;
        private byte[] wipeKey;
        private byte[] supplyKey;
        private boolean freezeDefault;
        private boolean kycDefault;
    }
```

-   Update `EntityTypeEnum` with `Token` type
```java
    public enum EntityTypeEnum {

        ACCOUNT(1),
        CONTRACT(2),
        FILE(3),
        TOPIC(4),
        TOKEN(5);

        private final int id;
    }
```

### Balance Persistence

> -   Update `AccountBalanceLineParser.parse` to parse additional token columns representing `TokenRelationships`.

Author: ```suggestion
-   Update `AccountBalanceLineParser.parse` to parse additional token column representing `TokenRelationships`.
```

-   Add `INSERT_TOKEN_BALANCE_STATEMENT` in `AccountBalancesFileLoader` to persist each `TokenRelationship`
```java
    private static final String INSERT_TOKEN_BALANCE_STATEMENT = "insert into token_balance " +
                "(consensus_timestamp, account_id, token_id, balance, symbol) values (?, ?, ?, ?, ?) on conflict do " +
                "nothing;";
```
-   Add `UPDATE_TOKEN_STATUS` in `AccountBalancesFileLoader` to update changes to `kycStatus` and `freezeStatus`
```java
    private static final String UPDATE_TOKEN_STATUS = "update token_balance set kyc_status = ?, " +
                "freeze_status = ? where token_id = ?";
```

> 

Author: Add TransactionHandlers
```suggestion
### Transaction Handlers
```

Author: Working on this

        entityListener.onEntityId(account);
        entityListener.onCryptoTransfer(new CryptoTransfer(consensusTimestamp, aa.getAmount(), account, tokenId, symbol));
    }
```
-   Add `insertTokenTransferList()` to handle `TokenTransferList` returned by `txRecord.getTokenTransferListsList()`
-   Update `insertTransferList(...)` to utilize `insertTransfer` and set defaults for tokenId and symbol

> _Note 1:_ For improved API calls it would be valuable to insert the symbol for cryptoTransfers. This is only available on the tokenCreate transaction, so a caching mechanism can be explored ot make this available for transfers

> _Note 2:_ There's an opportunity to refactor the `OnItem()` to focus better on different TransactionBody types but not necessarily within scope

## REST API
To achieve the goals and for easy integration with existing users the REST API should be updated in the following order
1.  The `accounts` REST API must be updated to support `tokenBalances`
2.  The `balances` REST API must be updated to support `tokenBalances`
3.  The `transactions` REST API must be updated to support `tokenTransfers`

> 4.  Add a Token Supply distribution REST API to show token distribution across accounts

Author: Add api url here for context

> 5.  Add a Token Discovery REST API to show available tokens on the network

Author: Eventually we'll need a tokenInfo endpoint
```suggestion
6.  Add a Token Info REST API to show details for a token on the network - `/api/v1/tokens/<symbol>`
```

-   Update `createTransferLists()` in `transactions.js` to build a `tokenTransfers` list if `symbol` isn't HBAR or if `token_id` is empty

### Token Supply distribution

`/api/v1/tokens/<symbol>/balances` this could be the equivalent of `/api/v1/balances?symbol=<symbol>` currently and would return a list of account and the token balances
```json
    {
        "timestamp": "0.000002345",
        "tokenBalances": [
          {
            "account": "0.15.10",
            "balance": 100
          },
          {
            "account": "0.15.9",
            "balance": 90
          },
          {
            "account": "0.15.8",
            "balance": 80
          }
        ],
        "links": {
          "next": null
        }
    }
```

To achieve this similar to balances
-   Create a `tokenBalance.js`
-   Add a `getTokenBalances()` which based on the time range provided (default to latest timestamp) and token `symbol` pulls out all the  `balance` and `symbol` values and returns them as an element in `tokenBalances` array

```sql
    select account_id, balance from token_balance where symbol = ? and consensus_timestamp = (select max(consensus_timestamp) from token_balance);
```

> Query parameters should include

Author: Need to circle back, these were initial suggestions based on current API's

Author: Provide parity with /accounts and /balances APIs where applicable i.e. filtering on accountId, tokenId and publicKey

> Query parameters should include

Author: ```suggestion
Optional Filters
```

-   `/api/v1/tokens/<symbol>/balances?id=0.0.1000`
-   `/api/v1/tokens/<symbol>/balances?balance=gt:1000`
-   `/api/v1/tokens/<symbol>/balances?timestamp=1566562500.040961001`

### Token Discovery
`/api/v1/tokens` this would return all tokens present on the network

```json
    {
        "timestamp": "0.000002345",
        "tokens": [
          {
            "symbol": "FOOBAR",
            "tokenId": "0.15.10",
            "key": {
              "_type": "ProtobufEncoded",
              "key": "7b2233222c2233222c2233227d"
            }
          },
          {
            "symbol": "FOOCOIN",
            "tokenId": "0.15.10",
            "key": {
              "_type": "ProtobufEncoded",
              "key": "9c2233222c2233222c2233227d"
            }
          }
        ],
        "links": {
          "next": null
        }
    }
```

To achieve this
-   Add a `getTokens()` to `tokenBalance.js` which pull the unique `symbol`, `tokenId` from the `token` table and joins with the `t_entities` table for the matching admin_key

```sql
    select symbol, tokenId e.key from token t
          join t_entities e on t.token_id = e.id and e.fk_entity_type_id = ${utils.ENTITY_TYPE_TOKEN};
```

> Additions

Author: ```suggestion
Optional Filters
```

# Hedera Token Service Design

## Purpose

The Hedera Token Service (HTS) builds upon the Cryptocurrency Service to provide decentralized issuance of custom tokens on the Hedera Network.

> The behavior will be similar to that of the native HBAR token and as such the Mirror Node will persist token balances and transfer lists and support the retrieval of information through its API's.

Reviewer: nit: apostrophe not needed.
```suggestion
The behavior will be similar to that of the native HBAR token and as such the Mirror Node will persist token balances and transfer lists and support the retrieval of information through its APIs.
```

Author: Updated

This document highlights the architecture and design changes to be made to support HTS.

> Changes should be applied in order of the following order to support easy 3rd integration.

Author: Updated

1.  Database Schema Updates
2.  Importer Updates for Ingestion - Token balance and token transactions
3.  Existing REST API Updates - Accounts, Balances and Transactions
4.  Token REST API additions
5.  Protobuf update for TokenService
6.  GRPC Token Service addition

## Goals
-   Ingest HTS Token transactions from record stream rcd files from the mainnet and persist information to the database

> -   Ingest account token balances from balance stream  CSV files from the mainnet and persist to the database

Reviewer: nit: extra space after stream
```suggestion
-   Ingest account token balances from balance stream CSV files from the mainnet and persist to the database
```

Author: Updated

-   Expose token balance of accounts in existing REST APIs that show balance
-   Provide a HTS REST API to return all tokens (Token Discovery)
-   Provide a HTS REST API to return all accounts holding a specific token (Token Supply Distribution)
-   Provide a streaming GRPC API to subscribe to token transfers for an account

## Non Goals

> -   Provide a streaming GRPC API to subscribe to HTS transfers for all accounts and or all tokens

Reviewer: nit: and/or
```suggestion
-   Provide a streaming GRPC API to subscribe to HTS transfers for all accounts and/or all tokens
```

Author: Updated

## Architecture

> ![Architecture](./images/mirror-arch.png)

Author: Renamed

1. Downloader retrieves transactions and balances (HBAR & Tokens) from stream files in cloud bucket and validates them
2. Parser persists to database
3. Client queries balance or token APIs for details
4. REST API retrieves token details and balances from database
5. Client subscribes to GRPC API for token transfers
6. GRPC API retrieves token transfers from database and returns to client
7. GRPC API is notified of new token transfers via database streaming logic

## Alternatives

## Database
To support the goals the following database schema changes should be made
1. `account_balance` table updated to use `account_id` of `entityId` type instead of `account_realm_num` and `account_num`
2. `t_entities` and ` t_entity_types` updated to represent new `TOKEN` entity
3. New `token_balance` table added to persist token only balances
4. New `token` table added to capture token specific entity properties

> 5. `crypto_transfer` table updated to persist token transfers and also distinguish between HBAR and other token transfers

Reviewer: I think that would also get rid of the need to allow nullable for the token_id when dealing with hbar, which would be nice.

> 5. `crypto_transfer` table updated to persist token transfers and also distinguish between HBAR and other token transfers

Author: Yeah, agree that might be a better approach, especially for the streaming reasons as tokens bubble out in size and transfers.
Will make this change

> 5. `crypto_transfer` table updated to persist token transfers and also distinguish between HBAR and other token transfers

Author: Added `token_transfer`

6. `t_transaction_types` and `t_transaction_results` updated to handle token transactions from network.
7.  New `token_account_info` table added to persist token specific account state

### Account Balance
```sql

>     alter table if exists account_balance

Author: Will remove from here as address already in #1070

        add column account_id entity_id null;

    update account_balance
        set account_id = encodeEntityId(0, account_realm_num, account_num);

    alter table if exists account_balance
        alter column account_id set not null;
```

### Crypto Transfer
-   Add `token_id` column to `crypto_transfer` table and allow nullable for hbar case which is not an entity
```sql
    alter table if exists crypto_transfer
        add column token_id entity_id null;
```

### Token Balance
-   Create `token_balance` table to distinctly capture token balances vs HBAR balances

```sql
    --- Add token_balance table to capture token account balances

>     create table if not exists token_balance

Author: Updated to just list out

>         consensus_timestamp bigint              primary key not null

Author: Updated

        account_id          entity_id           not null,

>         balance             bigint              not null

Author: Updated

        token_id            entity_id           not null;
```

### Token Account Info
-   Create `token_account_info` table to distinctly capture token account metadata changes

```sql
    --- Add token_account_info table to capture token-account info such as frozen, kyc and wipe status

>     create table if not exists token_account_info

Reviewer: I agree, this is the only table I'm having some trouble following.

>     create table if not exists token_account_info

Author: I don't think that's correct.
Freeze, KYC and Wipe are all operations that are performed on an account with respect to a particular account.
Here's more context.
1. You have an existing account w HBAR's or create a new account and fund it HBARs. Said account has no tokens
2. Token FOOBAR is created and supplied etc
3. [NEW] There's will be an `accept token` flag set for an account to accept FOOBAR that needs to be set to receive that token
4. Account is sent FOOBAR tokens. This adds the account to the ledger for this token with a nonzero balance. The account will have a freeze status based on the freezeDefault value for the token. It will have a wipe value set to false at beginning.
5. [NEW] KYC is likely needed for many tokens and may even be a precursor to receiving tokens. Need for KYC is now specified by the presence of a kycKey. If present then KYC must be performed on an account for a token before it can receive tokens. KYC for an account for a given token can be undone. In which case we need to be able to represent the toggling of the flag KYC
6. Similarly an account can be frozen and unfrozen in respect to a single token. i.e. You could have FOO tokens and BAR tokens and only have your account frozen for BAR tokens. If an account is frozen by default/toggled it needs to be unfrozen before tokens can be received.
7. [NEW] Wipe seems to have been cut off. Originally, an account for a given token could be wiped. This would also burn the tokens the account held for that token. Again, this would be done for only a given token and not all tokens and may or may not affect your actual account with respect to HBARs.

I've highlighted the new information I've received.
Ultimately the need for this table is to represent the different token->account properties that can change and are not related to HBAR.
If an account had 3 tokens they'd be represent by 3 rows here. Each row per token-account

>     create table if not exists token_account_info

Author: 2 new transaction types to be added to cover number 3.
These will mark the `onboarding` of an account to accept a token or to stop accepting.
At the time of accepting the defaults would be set and an entry in `token_account_info` would be added.
Future `TOKENFREEZE`, `TOKENUNFREEZE`,  `TOKENGRANTKYC`, `TOKENREVOKEKYC` and `TOKENWIPE` transactions would modify a single row in the `token_account_info` table
`TOKENASSOCIATE` & `TOKENDISSOCIATE`, yet to be added on service side

>     create table if not exists token_account_info

Reviewer: This makes more sense to me now, with those being per-account per-token operations I see the need for this.

>     create table if not exists token_account_info

Author: Updated to just list

        consensus_timestamp bigint              primary key,
        account_id          entity_id           not null,
        frozen              boolean             not null,
        kyc                 boolean             not null,
        token_id            entity_id           not null,
        wiped               boolean             not null default false;
```

> _Note:_  `frozen` and `kyc` are set by `TokenCreation.freezeDefault` and `TokenCreation.kycDefault` respectively

### Entity Types
-   Add new `t_entity_types` row with name `token`
```sql
    insert into t_entity_types (id, name) values (5, 'token');
```

### Transaction Types
-   Add new `t_transaction_types`:
```sql
    insert into t_transaction_types (proto_id, name) values

>         (56, 'TOKENCREATE'),

Author: Updated

        (57, 'TOKENTRANSACT'),

>         (58, 'TOKENGETINFO'),

Author: Updated

>         (59, 'TOKENFREEZEACCOUNT'),

Author: Updated

        (177, 'INVALID_TOKEN_REF'),
        (178, 'ACCOUNT_KYC_NOT_GRANTED_FOR_TOKEN'),
        (179, 'TOKEN_HAS_NO_KYC_KEY'),
        (180, 'INSUFFICIENT_TOKEN_BALANCE'),
        (181, 'TOKEN_WAS_DELETED'),
        (182, 'TOKEN_HAS_NO_SUPPLY_KEY'),
        (183, 'TOKEN_HAS_NO_WIPE_KEY'),
        (184, 'INVALID_TOKEN_MINT_AMOUNT'),
        (185, 'INVALID_TOKEN_BURN_AMOUNT'),
        (186, 'ACCOUNT_HAS_NO_TOKEN_RELATIONSHIP')
        (187, 'CANNOT_WIPE_TOKEN_TREASURY_ACCOUNT')
        (188, 'INVALID_KYC_KEY')
        (189, 'INVALID_WIPE_KEY')
        (190, 'INVALID_FREEZE_KEY')
        (191, 'INVALID_SUPPLY_KEY')
        (192, 'INVALID_TOKEN_EXPIRY')
        (193, 'TOKEN_HAS_EXPIRED')
        (194, 'TOKEN_IS_IMMUTABlE');
```

### Entities
-   Insert token entity info into `t_entities` table from `TokenCreation.proto` transaction object. Insert `adminKey` as `key`, `expiry` as `exp_time_ns`, `autoRenewAccount` as `auto_renew_account_id` and `autoRenewPeriod` as `auto_renew_period`.

### Token
-   Create `token` table to split out the non shared entity items into a new table. Most API calls may not require this information and therefore additional sql joins may be avoided.
```sql

>     create table if not exists token

Author: Updated as listing

        token_id            entity_id               primary key,
        consensus_timestamp bigint                  not null,
        divisibility        bigint                  not null,
        freeze_default      boolean                 not null,
        freeze_key          bytea,
        initial_supply      bigint                  not null,
        kyc_default         boolean                 not null,
        kyc_key             bytea,

>         supply_key          bytea,

Author: I think it can be null. IN the doc there's expected behavior for when it is null under the "Key Types" section

>         supply_key          bytea,

Reviewer 3: yes you are right

>         token               character varying(96)   not null,

Author: Yes will revert back to `symbol` and I'm adding a serial primary key.
There will be 2 indexes added
```
create unique index if not exists token_account__token_account on token_account(token_id, account_id);

create index if not exists token_account__token_account_timestamp
     on token_account (token_id desc, account_id desc, create_timestamp desc);
```

```java
    public class TokenBalance implements Persistable<TokenBalance.Id> {

        private long balance;

        @EmbeddedId
        @JsonUnwrapped
        private Id id;

        @Override
        public boolean isNew() {
            return true; // Since we never update balances and use a natural ID, avoid Hibernate querying before insert
        }

        @Data
        @AllArgsConstructor
        @NoArgsConstructor
        @Embeddable
        public static class Id implements Serializable {

            private static final long serialVersionUID = 1416544357338674260L;

            private long consensusTimestamp;

            @Convert(converter = TokenIdConverter.class)
            @JsonSerialize(using = EntityIdSerializer.class)
            private EntityId tokenId;

            @Convert(converter = AccountIdConverter.class)
            @JsonSerialize(using = EntityIdSerializer.class)
            private EntityId accountId;
        }
    }
```

-   Add `Token` class to hold Token specific metadata outside of the base entity
```java
    public class Token {
        private EntityId tokenId;
        private String token;
        private Long initialSupply;

>         private Long  divisibility;

Author: Ended up removing class member details so as not to be over descriptive.
However intiialSupply should be fine as a `long` no? Then `divisibility` can be reduced to an `int`

        @Convert(converter = AccountIdConverter.class)
        private EntityId treasuryId;
        private byte[] kycKey;
        private byte[] freezeKey;
        private byte[] wipeKey;
        private byte[] supplyKey;
        private boolean freezeDefault;
        private boolean kycDefault;
    }
```

### Balance Parsing

> To support HTS the balance CSV adds a version comment to the file, comments the Timestamp and adds a `accountMetadata` column which is a binary representation of `TokenRelationship`.

Author: Updated

To allow the mirror node to support both V1 and V2 balance files
-   Rename `BalanceFileReaderImpl` to `BalanceFileReaderImplV1`
-   Add a `BalanceFileReaderImplV2` class to parse `shardNum,realmNum,accountNum,balance,accountMetadata`.  Reader should set `tokenBalances` in `AccountBalance`.
-   Add a `CompositeBalanceFileReader` that implements `BalanceFileReader` and reads the first few headers lines of the CSV
    -   If the first line is TimeStamp use `BalanceFileReaderImplV1` as the reader

>     -   If the first line is `# version:2`, and 2nd line is `# TimeStamp` then use `BalanceFileReaderImplV2` as the reader

Author: Updated

    -   If no `version` and `Timestamp` found throw an `InvalidDatasetException`
-   Add `INSERT_TOKEN_BALANCE_STATEMENT` in `AccountBalancesFileLoader` to persist each `TokenRelationship`
```java

>     private static final String INSERT_TOKEN_BALANCE_STATEMENT = "insert into token_balance " +

Author: Removed

### Transaction Handlers
Additional handlers will be needed to support the added Token transaction types.
In all cases override `getEntity()`, pulling entity info from appropriate transactionBody

-   Add `TokenCreateTransactionsHandler`
    -   override `updatesEntity()` to return true
-   Add `TokenTransactTransactionsHandler`
-   Add `TokenFreezeAccountTransactionsHandler`
-   Add `TokenUnfreezeAccountTransactionsHandler`
-   Add `TokenGrantKycToAccountTransactionsHandler`
-   Add `TokenRevokeKycFromAccountTransactionsHandler`
-   Add `TokenDeleteTransactionsHandler`
    -   override `updatesEntity()` to return true
-   Add `TokenUpdateTransactionsHandler`
    -   override `updatesEntity()` to return true
-   Add `TokenMintTransactionsHandler`
-   Add `TokenBurnTransactionsHandler`
-   Add `TokenAccountWipeTransactionsHandler`

### Token Transfer Parsing

Modify `EntityRecordItemListener` to handle parsing HTS transactions

-   Modify `OnItem()` to check for `TransactionBody.hasTokenCreation()`

> -   Add `insertTokenCreateTransferList()` and `insertTokenTransferList()` to parse out `txRecord.getTokenTransferListsList()`. Create a new CryptoTransfer object for each `TokenTransferList` item and pass it to `entityListener.onCryptoTransfer()`. Also, pull out `tokenID` from `TokenTransferList.token`

Author: Simplified

-   Add `insertTransfer()` that can be shared for `TransferList` and `TokenTransferList`
```java
    private void insertTransfer(long consensusTimestamp, AccountAmount aa, EntityId account, EntityId tokenId) {
        entityListener.onEntityId(account);
        entityListener.onCryptoTransfer(new CryptoTransfer(consensusTimestamp, aa.getAmount(), account, tokenId));
    }
```
-   Add `insertTokenTransferList()` to handle `TokenTransferList` returned by `txRecord.getTokenTransferListsList()`
-   Update `insertTransferList(...)` to utilize `insertTransfer` and set defaults for `tokenId` and `token`

## REST API
To achieve the goals and for easy integration with existing users the REST API should be updated in the following order

> 1.  The `accounts` REST API must be updated to support `tokenBalances`

Author: Updated

2.  The `balances` REST API must be updated to support `tokenBalances`
3.  The `transactions` REST API must be updated to support `tokenTransfers`
4.  Add a Token Supply distribution REST API to show token distribution across accounts - `/api/v1/tokens/<token>/balances`
5.  Add a Token Discovery REST API to show available tokens on the network - `/api/v1/tokens`
6.  Add a Token Info REST API to show details for a token on the network - `/api/v1/tokens/<token>`

### Accounts Endpoint
-   Update `/api/v1/accounts` response to add token balances
```json
{
    "accounts": [
      {
        "balance": {
          "timestamp": "0.000002345",
          "balance": 80,

>           "tokenBalances": [

Reviewer 2: `tokens`

Author: Updated

Reviewer 2: This doesn't seem to have changed.

Author: Updated

            {
              "token": "FOOBAR",
              "balance": 80
            },
            {
              "token": "FOOCOIN",
              "balance": 50
            }
          ]
        },
        "account": "0.0.8",
        "expiry_timestamp": null,
        "auto_renew_period": null,
        "key": null,
        "deleted": false
      }
    ],
    "links": {
      "next": null
    }
  }
```

> To achieve this

Author: Removed

-   Update `accounts.js` `getAccountss()` to add an additional join to pull the `balance` column where `account_id` matches between tables `token_balance` and `t_entities`, assign each row to an element of `tokenBalances`. `token` can be retrieved with additional join from `token` table or by a caching mechanism for `tokenId` -> `token` mapping

### Balances Endpoint
-   Update `/api/v1/balances` response to add token balances
```json
    {
        "timestamp": "0.000002345",
        "balances": [
          {
            "account": "0.0.8",
            "balance": 100,

>             "tokenBalances": []

Author: Updated

>             "tokenBalances": []

Reviewer 2: This doesn't seem to have changed.

>             "tokenBalances": []

Author: Updated

          },
          {
            "account": "0.0.10",
            "balance": 100,
            "tokenBalances": [
              {

>                 "token": "FOOBAR",

Author: This is certainly more succinct.
However, this would be the first section in all of our current REST API responses where there's no label for a value.
So it breaks the format in that sense.

### Transactions Endpoint
-   Update `/api/v1/transactions` response to add token transfers
```json
    {
      "transactions": [
        {
          "consensus_timestamp": "1234567890.000000001",
          "valid_start_timestamp": "1234567890.000000000",
          "charged_tx_fee": 7,
          "memo_base64": null,
          "result": "SUCCESS",
          "transaction_hash": "aGFzaA==",
          "name": "CRYPTOTRANSFER",
          "node": "0.0.3",
          "transaction_id": "0.0.10-1234567890-000000000",
          "valid_duration_seconds": "11",
          "max_fee": "33",
          "transfers": [
            {
              "account": "0.0.9",
              "amount": 10
            },
            {
              "account": "0.0.10",
              "amount": -11
            },
            {
              "account": "0.0.98",
              "amount": 1
            }
          ],
          "token_transfers": [
            {

>               "token":"FOOBAR",

Author: Yeah, I'm okay with managing duplication if we get more simplicity and matching db structure.
Will update

>               "token":"FOOBAR",

Author: Yeah looks like `symbol` for now and then the `/api/v1/tokens` can be where we return `symbol` and/or `tokenId`

>               "token":"FOOBAR",

Author: Actually looks like there's still some discussion on this, will circle back

>               "token":"FOOBAR",

Author: Final call is `tokenId` everywhere

          {
            "account": "0.15.10",
            "balance": 100
          },
          {
            "account": "0.15.9",
            "balance": 90
          },
          {
            "account": "0.15.8",
            "balance": 80
          }
        ],
        "links": {
          "next": null
        }
    }
```

To achieve this similar to balances
-   Create a `tokenBalance.js`
-   Add a `getTokenBalances()` which based on the time range provided (default to latest timestamp) and `tokenId` pulls out `account_id` and `balance` values from `token_balance` and returns them as an element in `tokenBalances` array.

```sql
    select account_id, balance from token_balance where token_id = ? and consensus_timestamp = (select max(consensus_timestamp) from token_balance);
```

Optional Filters
-   `/api/v1/tokens/<token>/balances?id=0.0.1000`
-   `/api/v1/tokens/<token>/balances?balance=gt:1000`
-   `/api/v1/tokens/<token>/balances?timestamp=1566562500.040961001`

### Token Discovery
`/api/v1/tokens` this would return all tokens present on the network

```json
    {
        "timestamp": "0.000002345",
        "tokens": [
          {
            "token": "FOOBAR",

>             "tokenId": "0.15.10",

Author: Updated

>             "key": {

Reviewer 3: should this be `adminKey`?

Author: Yes, but I believe other REST APIs were using just key so I tried to match. Will confirm

Author: I can change it to be clear though since tokens do have many keys

>             "key": {

Author: Design request seemed to want admin key as part of discoverability, can circle back

>             "key": {

Author: Yep per design request

> A client application can call an API to return all tokens issued 
> Returns token name, symbol, and admin key 
> Can also be specified by admin key(s)

              "_type": "ProtobufEncoded",
              "key": "7b2233222c2233222c2233227d"
            }
          },
          {
            "token": "FOOCOIN",
            "tokenId": "0.15.11",
            "key": {
              "_type": "ProtobufEncoded",
              "key": "9c2233222c2233222c2233227d"
            }
          }
        ],
        "links": {
          "next": null
        }
    }
```

To achieve this
-   Add a `getTokens()` to `tokenBalance.js` which pull the unique `token`, `tokenId` from the `token` table and joins with the `t_entities` table for the matching admin_key

```sql
    select token, token_id from token t
          join t_entities e on t.token_id = e.id and e.fk_entity_type_id = ${utils.ENTITY_TYPE_TOKEN};
```

Optional Filters

> -   `/api/v1/tokens?adminkey=HJ^&8` - All tokens with matching admin key

Reviewer 3: protobuf `Key` is a complicated structure, it can be a single key of different algorithms, it can be a threshold key with a threshold then recursively includes `Key`, or a `KeyList` recursively includes a list of `Key`s

in order to support such a filter, we need more clarity on how it works

Author: The example was a placeholder.
More realistic query is exactly what we have in /api/v1/accounts
```suggestion
-   `/api/v1/tokens?publickey=3c3d546321ff6f63d701d2ec5c277095874e19f4a235bee1e6bb19258bf362be` - All tokens with matching admin key
```

Author: The t_entities.key is also from the protobuf `key`, we're currently able to filter on that for the /api/v1/accounts?account.publickey=<key>.
So I believe it's not as complex as we think. It's also the exact same key so the sql query logic is the same and should be doable

-   `/api/v1/tokens?account.id=0.0.8` - All tokens for matching account

### Token Info
-   Add a `getTokenInfo()` to `tokens.js` to retrieve token info from `token` table
```json
    {
      "token": "FOOCOIN",
      "tokenId": "0.15.10",
      "treasury_account": "0.15.10",
      "adminKey": {
        "_type": "ProtobufEncoded",
        "key": "9c2233222c2233222c2233227d"
      },
      "kycKey": {
        "_type": "ProtobufEncoded",
        "key": "9c2233222c2233222c2233227d"
      },
      "freezeKey": {
        "_type": "ProtobufEncoded",
        "key": "9c2233222c2233222c2233227d"
      },
      "wipeKey": {
        "_type": "ProtobufEncoded",
        "key": "9c2233222c2233222c2233227d"
      },
      "supplyKey": {
        "_type": "ProtobufEncoded",
        "key": "9c2233222c2233222c2233227d"
      },
      "freezeDefault": false,
      "kycDefault": false,
      "expiryTimestamp": null,
      "autoRenewAccount": "0.0.6",
      "autoRenewPeriod": null
    }
```

## Protobuf

```proto
    message TokenQuery {

>         .proto.TokenID tokenID = 1; // The token ID to retrieve transfers for

Reviewer 3: may want to support `TokenRef` instead of just `TokenID`

Author: Yes good idea

Author: Noted locally but removed those details in the section

```java
    package com.hedera.mirror.grpc.service;

    public class TokenServiceImpl extends AbstractServiceImpl implements TokenService {
        ...
        private final Retriever retriever;
        Flux<StreamItem> subscribeTokenTransfers(@Valid StreamItemFilter filter){
            ...
        }

        ...
    }
```

## Non-Functional Requirements

## Open Questions
-   [x] What's the maximum character size of the token `symbol` string
    -   A: Max length is currently 32 chars https://github.com/hashgraph/hedera-services/blob/master/hedera-node/src/main/resources/bootstrap.properties#L52
-   [x] Will a `token_id` and `token` be assigned a default value for HBARs across the network e.g. i.e. '1' and 'HBAR' respectively
    -   A: Currently no since hbar is not treated as an entity like tokens will be.
-   [ ] Should token only entity items exist in their own table or be added to `t_entities`?
-   [x] Should `account_balance` `accountNum` and `accountRealmNum` be migrated into `entityId` or should token_balance also use `accountNum` and `accountRealmNum` instead of `entityId`?
    -   A: Should be migrated to use `account_id` only

> -   [ ] What filer options should be provided for new Token API's

Author: Updated
